os/fs/driver/mtd/smart.c

```diff
@@ -5109,13 +5147,169 @@ static int smart_joutnal_set_area_state(FAR struct smart_struct_s *dev, int area
 	ret = MTD_WRITE(dev->mtd, readaddress, 1, &id_bits);
 	if (ret != 1) {
 		fdbg("Error updating physical sector %d status\n", psector);
-		return -EIO;
+		return ret;
 	}
 
 	dev->journal_area_state[area] = id_bits;
 	return OK;
 }
 
+/****************************************************************************
+ * Name: smart_journal_get_writeaddress
+ *
+ * Description:
+ *   Get address to write next journal data
+ *
+ ****************************************************************************/
+
+static int smart_journal_get_writeaddress(FAR struct smart_struct_s *dev, uint32_t *address)
+{
+	uint16_t psector;
+	size_t seq;
+	uint32_t *offset = address;
+
+	psector = dev->neraseblocks * dev->sectorsPerBlk;
+
+	/* Get Physical sector & sequence of each area First */
+	if (dev->area == 0) {
+		if (dev->journal_seq >= dev->njournaldata / 2) {
+			fdbg("journal sequence & area bug!! sequence : %d\n", dev->journal_seq);
+			return -EFAULT;
+		}
+		psector = dev->neraseblocks * dev->sectorsPerBlk;
+		seq = dev->journal_seq;
+	} else {
+		if (dev->journal_seq < dev->njournaldata / 2) {
+			fdbg("journal sequence & area bug!! sequence : %d\n", dev->journal_seq);
+			return -EFAULT;
+		}
+		psector = dev->neraseblocks * dev->sectorsPerBlk + (dev->njournalsectors / 2);
+		seq = dev->journal_seq - (dev->njournaldata / 2);
+	}
+
+	*offset = (psector * dev->sectorsize) + (seq * sizeof(journal_log_t) + 1);
+	fvdbg("address : %d psector : %d seq : %d dev seq : %d\n", offset, psector, seq, dev->journal_seq);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: smart_journal_checkin
+ *
+ * Description:
+ *   Check in journal data
+ *
+ ****************************************************************************/
+
+static int smart_journal_checkin(FAR struct smart_struct_s *dev, journal_log_t *log, uint32_t address)
+{
+	int ret;
+	journal_log_t buf;
+	size_t log_size = sizeof(journal_log_t);
+
+	/* First write Journal log with Checkin state */
+	ret = MTD_WRITE(dev->mtd, address, log_size, (FAR uint8_t *)log);
+	if (ret != log_size) {
+		fdbg("Checkin Error, write log data... physical address %u ret : %d\n", address, ret);
+		return ret;
+	}
+
+	/* Verify written data */
+	ret = MTD_READ(dev->mtd, address, log_size, (FAR uint8_t *)&buf);
+	if (ret != log_size) {
+		fdbg("Checkin Error, read log data... physical address %u ret : %d\n", address, ret);
+		return ret;
+	}
+
+	if (UINT8TOUINT16(buf.crc16) != smart_calc_journal_crc(&buf)) {
+		fdbg("Checkin Error, verify crc failed buf crc16 : %d calc crc : %d\n", UINT8TOUINT16(buf.crc16), smart_calc_journal_crc(&buf));
+		return ret;
```

Comment:
This error type should be discriminated from the error, which might happen at MTD_READ().

Reply:
Thanks. I will apply it